include/zephyr/drivers/pwm.h

```diff
@@ -400,6 +419,52 @@ typedef void (*pwm_capture_callback_handler_t)(const struct device *dev,
 					       uint32_t pulse_cycles,
 					       int status, void *user_data);
 
+struct pwm_event_callback;
+
+/**
+ * @brief PWM event callback handler function signature.
+ *
+ * @note The callback handler can be called in an interrupt context.
+ *
+ * @note @kconfig{CONFIG_PWM_EVENT} must be selected to enable PWM event
+ * support.
+ *
+ * @param[in] dev PWM device instance.
+ * @param callback Original struct gpio_callback owning this handler.
+ * @param channel PWM channel.
+ * @param events Event mask. See @ref PWM_EVENT_FLAGS.
+ *
+ */
+typedef void (*pwm_event_callback_handler_t)(const struct device *dev,
+					     struct pwm_event_callback *callback, uint32_t channel,
+					     pwm_flags_t events);
+
+/**
+ * @brief PWM event callback structure
+ *
+ * Used to register an event callback in the driver instance callback list.
+ * As many callbacks as needed can be added as long as each of them
+ * are unique pointers of struct pwm_event_callback.
+ * Beware such structure should not be allocated on stack.
+ *
+ * Note: to help setting it, see pwm_init_event_callback().
+ *
+ */
+struct pwm_event_callback {
+	/** @cond INTERNAL_HIDDEN */
+	sys_snode_t node;
+	/** @endcond */
+
+	/** Actual callback function being called when relevant. */
+	pwm_event_callback_handler_t handler;
+
+	/** Channel the callback is interested in. */
+	uint32_t channel;
+
+	/** A mask of events the callback is interested in. */
+	pwm_flags_t event_mask;
```

Comment:
I don't think we should be reusing the `pwm_flags_t` type for storing PWM events types. The flags are for channel configuration, not for event masks etc.